voila/handler.py

```diff
@@ -51,7 +51,20 @@ def get(self, path=None):
             raise tornado.web.HTTPError(404, 'file not found')
 
         # Fetch kernel name from the notebook metadata
-        kernel_name = notebook.metadata.get('kernelspec', {}).get('name', self.kernel_manager.default_kernel_name)
+        kernelspec = notebook.metadata.get('kernelspec', {})
+        kernel_name = kernelspec.get('name', self.kernel_manager.default_kernel_name)
+
+        # Find a spec matching the language if the kernel name does not exist in the kernelspecs
+        all_kernel_specs = yield tornado.gen.maybe_future(self.kernel_spec_manager.get_all_specs())
+        if kernel_name not in all_kernel_specs:
+            kernel_language = kernelspec.get('language', 'python').lower()
+            matches = [
+                name for name, kernel in all_kernel_specs.items()
+                if kernel["spec"]["language"].lower() == kernel_language
+            ]
+            if matches:
+                matches.sort(key=lambda name: all_kernel_specs[name]["spec"]["display_name"])
+                kernel_name = matches[0]
```

Comment:
I wonder if we have 0 matches, if we can have a custom template with a proper error msg and http error code: https://en.wikipedia.org/wiki/List_of_HTTP_status_codes

If there are 1 or more matches, I think we should log a warning, so users are aware of it (I can imagine someone expecting it to run with a conda-env-kernel, which cannot be found, leading to surprises).

Reply:
Agreed, a warning would definitely be useful to let the users know the kernel being used is not the one they expect.

Having a custom template with a proper error message is also a good idea, but maybe for a follow-up PR?